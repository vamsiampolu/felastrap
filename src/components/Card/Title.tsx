import { useFela } from "react-fela";
import { Header, type HeaderProps } from "../Header/Header";
import { StyleProps } from "../common";

export type CardTitleProps = HeaderProps & { color?: string };

const rule = ({ color }: StyleProps<CardTitleProps>) => {
  return {
    color,
    marginBottom: "1rem",
  };
};

export const CardTitle = ({
  as = "h5",
  className = "",
  color = "#000",
  children,
  ...other
}: CardTitleProps) => {
  const { css } = useFela({ color });
  const cx = `${css(rule)} ${className || ""}`.trim();

  return (
    <Header data-testid="card-title" as={as} className={cx} {...other}>
      {children}
    </Header>
  );
};
