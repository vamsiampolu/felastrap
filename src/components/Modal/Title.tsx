import { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";

const rule = () => ({
  margin: 0,
  lineHeight: 1.42857143,
  textAlign: "center",
});

export type ModalTitleProps = ComponentPropsWithoutRef<"h4"> & {
  title: string;
};

export const ModalTitle = ({
  className = "",
  title,
  ...props
}: ModalTitleProps) => {
  const { css } = useFela();
  const cx = `${css(rule)} bs-modal-title ${className}`.trim();
  return (
    <h4 data-testid="modal-title" className={cx} {...props}>
      {title}
    </h4>
  );
};
