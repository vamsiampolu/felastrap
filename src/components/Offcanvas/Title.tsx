import { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";

const rule = () => ({
  marginBottom: 0,
  lineHeight: 1.5,
});

export type OffcanvasTitleProps = ComponentPropsWithoutRef<"h5">;

export const OffcanvasTitle = ({
  className = "",
  id,
  children,
}: OffcanvasTitleProps) => {
  const { css } = useFela();
  const cx = `${css(rule)} offcanvas-title ${className}`.trim();
  return (
    <h5 data-testid="offcanvas-title" className={cx} id={id}>
      {children}
    </h5>
  );
};
