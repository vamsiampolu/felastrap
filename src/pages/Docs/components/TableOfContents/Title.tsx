import { ReactNode } from "react";
import { useFela } from "react-fela";

const strongRule = {
  display: "none",
  "@media (min-width: 768px)": {
    display: "block",
  },
  marginLeft: "1rem",
  marginTop: "0.5rem",
  marginBottom: "0.5rem",
};

export const Title = ({ children }: { children: ReactNode }) => {
  const { css } = useFela();
  return <strong className={css(strongRule)}>{children}</strong>;
};
