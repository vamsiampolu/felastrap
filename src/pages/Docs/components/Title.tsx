import { PropsWithChildren } from "react";
import { Header } from "../../../components/Header/Header";
import { AnchorLink } from "./AnchorLink";

export type TitleProps = PropsWithChildren<{
  id: string;
  "aria-label": string;
  small?: boolean;
}>;

export const Title = (props: TitleProps) => {
  const { id, children: label, small = false } = props;
  const as = small ? "h3" : "h2";
  const ariaLabel = props["aria-label"];
  return (
    <Header as={as} id={id}>
      {label} <AnchorLink href={`#${id}`} aria-label={ariaLabel} />
    </Header>
  );
};
