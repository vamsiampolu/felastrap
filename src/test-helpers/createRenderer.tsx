import expandShorthand from "fela-plugin-expand-shorthand";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";
import prefixer from "fela-plugin-prefixer";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import { createRenderer as createFelaRenderer, type IRenderer } from "fela";
import { lightTheme } from "../reboot/lightTheme";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  const enhancers = [sortClassNames(), sortMediaQueryMobileFirst()];
  const plugins = [
    prefixer(),
    placeholderPrefixer(),
    expandShorthand(autoMergeLonghands),
  ];

  return createFelaRenderer({
    plugins,
    enhancers,
  });
};

export const cssReboot = (renderer: IRenderer) => {
  renderer.renderStatic(lightTheme);
};
