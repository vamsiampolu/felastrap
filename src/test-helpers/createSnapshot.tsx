import { IRenderer } from "fela";
import { renderToString } from "fela-tools";
import { ReactElement, ReactNode } from "react";

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import {
  type UserEvent,
  type UserEventApi,
} from "@testing-library/user-event/dist/types/setup/setup";
import * as prettier from "prettier";
import { renderToStaticMarkup } from "react-dom/server";
import { Wrapper } from "./Wrapper";

async function formatCSS(css: string) {
  return prettier.format(css, {
    parser: "css",
    useTabs: false,
    tabWidth: 2,
  });
}

function formatHTML(str: string) {
  return prettier.format(str, { parser: "html" });
}

function renderToMarkup(component: ReactElement): string {
  return renderToStaticMarkup(component);
}

export const createSnapshot = async ({
  component,
  renderer,
  theme = {},
}: {
  component: ReactNode;
  renderer: IRenderer;
  theme?: object;
}) => {
  // reset renderer to have a clean setup
  renderer.clear();

  const element = (
    <Wrapper theme={theme} renderer={renderer}>
      {component}
    </Wrapper>
  );

  const view = renderToMarkup(element);
  const str = renderToString(renderer);

  const css = await formatCSS(str);
  const html = await formatHTML(view);

  return `${css}\n\n${html}`;
};

export type EventName = keyof UserEventApi | "focus" | "focus-visible";
export type CreateUserEventType = {
  eventName: EventName;
  component: ReactElement;
  renderer: IRenderer;
  testId: string;
  theme?: object;
  donotWrap?: boolean;
};

export type SetupEvent = {
  user: UserEvent;
  eventName: EventName;
  element: HTMLElement;
};
const setupEvent = async ({ user, eventName, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.hover(element);
      break;
    }

    case "focus": {
      element.focus();
      break;
    }

    case "focus-visible": {
      element.contentEditable = "true";
      element.focus();
      element.contentEditable = "false";
      break;
    }
  }
};

const clearEvent = async ({ eventName, user, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.unhover(element);
      break;
    }

    case "focus":
    case "focus-visible": {
      element.blur();
      break;
    }
  }
};

export const createUserEventSnapshot = async ({
  eventName,
  component,
  testId,
  renderer,
  theme = {},
  donotWrap = false,
}: CreateUserEventType) => {
  renderer.clear();

  const user = userEvent.setup();
  let wrappedComponent = component;
  if (!donotWrap) {
    wrappedComponent = (
      <Wrapper theme={theme} renderer={renderer}>
        {component}
      </Wrapper>
    );
  }

  render(wrappedComponent);

  const element = await screen.findByTestId(testId);

  setupEvent({ element, eventName, user });
  const view = await renderToMarkup(wrappedComponent);
  const utils = renderToString(renderer);
  const css = await formatCSS(utils);
  const html = await formatHTML(view);

  clearEvent({ eventName, element, user });

  return `${css}\n\n${html}`;
};
