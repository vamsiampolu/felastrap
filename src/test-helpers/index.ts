import { Wrapper } from "./Wrapper";
import { createRenderer, cssReboot } from "./createRenderer";
import { createSnapshot, createUserEventSnapshot } from "./createSnapshot";
import { resizeHeight, resizeWidth } from "./resize";

export {
  createRenderer,
  cssReboot,
  createSnapshot,
  createUserEventSnapshot,
  resizeWidth,
  resizeHeight,
  Wrapper,
};
