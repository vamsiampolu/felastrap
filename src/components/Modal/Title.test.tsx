import { ModalTitle } from "./Title";
import { IRenderer } from "fela";
import { createRenderer, createSnapshot, cssReboot } from "../../test-helpers";

describe("components/Modal/Title", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a Modal Title", async () => {
    const component = <ModalTitle title="Modal Title" />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
