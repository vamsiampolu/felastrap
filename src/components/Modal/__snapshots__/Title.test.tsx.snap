// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal/Title should render a Modal Title 1`] = `
".a {
  line-height: 1.42857143;
}
.b {
  text-align: center;
}
.c {
  margin-top: 0;
}
.d {
  margin-right: 0;
}
.e {
  margin-bottom: 0;
}
.f {
  margin-left: 0;
}


<h4 data-testid="modal-title" class="a b c d e f bs-modal-title">
  Modal Title
</h4>
;
"
`;
